### PR TITLE
release-job-migrator: initial commit

### PR DESCRIPTION
**Pull request by AlexNPavel**
This PR adds a new command to ci-tools that migrates release jobs from
hand-crafted jobs to jobs generated by a ci-operator config file. There
are 2 ways it can do this conversion:
1. If a job has an `UNRESOLVED_CONFIG` set, it can identify required base images and tests from there.
2. A config file can be used to configure jobs that use the same test name.

The config file is simply a yaml map where the key is the base test
name of a job (for example, for job `release-openshift-origin-installer-e2e-gcp-upgrade-4.7`,
the base test name is `gcp-upgrade`) and the value contains 2 things:
1. The corresponding multi-stage test configuration.
2. Any extra base images that the test may require.

This is an example of the config:
```
config:
  e2e-metal-ipi:
    steps:
      cluster_profile: packet
      env:
        DEVSCRIPTS_CONFIG: |
          IP_STACK=v4
          NETWORK_TYPE=OpenShiftSDN
      workflow: baremetalds-e2e
    base_images:
      dev-scripts:
        name: test
        namespace: openshift-kni
        tag: dev-scripts
```

The tool identifies what ci-operator config the tests and images should
be added to based on the name of the periodic job. If the ci-operator
config file does not exist yet, it creates a new one with the proper release
and base image set.

Currently the tool does not support handling of jobs that upgrade
between different release streams. That will be added in a future
PR.

### CODE REVIEW

File: cmd/release-job-migrator/main.go
Comment (stevekuznetsov) [inline cmd/release-job-migrator/main.go:90]: commented log lines :)

File: cmd/release-job-migrator/main.go
Comment (stevekuznetsov) [inline cmd/release-job-migrator/main.go:298]: note: the code handles this but IIRC there are jobs that dynamically edit their own config before invoking `ci-operator` for which the intent may have been to dynamically generate the config? in any case we will handle it later ...

**Review by stevekuznetsov — APPROVED** [review]
This looks really good!

/hold
Letting @petr-muller look

**Comment by petr-muller** [comment]
/hold

I'll definitely want to have a look today

File: cmd/release-job-migrator/main.go
Comment (AlexNPavel) [reply cmd/release-job-migrator/main.go:298]: This shouldn't be a problem for the jobs that this tool will handle. It's configured to only migrate jobs that use `ci-operator` as the command and simply print a warning notifying the user that that job will be ignored if not. (https://github.com/openshift/ci-tools/pull/1450/files#diff-fff22e1ebd0d1f2036370917bb8fc4706915281da9fcd2c6f2b15cdd36eabca2R264-R267)

**Review by petr-muller — COMMENTED** [review]
Neat job, mostly nits. It reads a little hacky, but it's a temporary single purpose tools so whatever :)

File: cmd/release-job-migrator/main.go
Comment (AlexNPavel) [reply cmd/release-job-migrator/main.go]: Ah, log lines from when I trying to debug an issue that was actually in a completely different function that I didn't expect. Removed.

File: test/integration/release-job-migrator/input/jobs/openshift-release-release-4.6-periodics.yaml
Comment (AlexNPavel) [reply test/integration/release-job-migrator/input/jobs/openshift-release-release-4.6-periodics.yaml:11]: Yes, the command tells the user to run `make update` on the release repo after it's done running, which runs prowgen

File: test/integration/release-job-migrator/expected.txt
Comment (AlexNPavel) [reply test/integration/release-job-migrator/expected.txt:9]: No, this is correct. Some jobs we won't be able to replace until we have a valid workflow for them (like aws-compact), so this is just listing what jobs which haven't been configured yet (either via the generator config or via UNRESOLVED_CONFIG)

File: cmd/release-job-migrator/main.go
Comment (AlexNPavel) [reply cmd/release-job-migrator/main.go:285]: Based on running grep for `--target` on the release-controller's jobs directory, all jobs do `--target=yada`

File: test/integration/release-job-migrator/expected.txt
Comment (petr-muller) [reply test/integration/release-job-migrator/expected.txt:9]: In that case I think the message could be a bit more verbose about the consequence...

File: cmd/release-job-migrator/main.go
Comment (AlexNPavel) [reply cmd/release-job-migrator/main.go:299]: If this is the case, we won't be able to really use `UNRESOLVED_CONFIG` much. All (or almost all) configs use the `$(BRANCH)` env var, and most seem to use `$(CLUSTER_TYPE)`. Maybe I could restrict the env var replacing to just `CLUSTER_TYPE` and `BRANCH`? `CLUSTER_TYPE` should always be a simple string matching our usual cluster types (aws, azure4, gcp, packet, ...) and `BRANCH` should always match `4.X`.

Comment (petr-muller) [reply cmd/release-job-migrator/main.go:298]: Yeah, after some thought, let's go with what you have right now. In this tool it's fine - it's the one-time migration and we'll immediately get a validation in a rehearsal. I came here from solving the same problem in `pj-rehearse` where the consequences of pathological cases are more obscure, so I thought this is a larger problem than it really is. 

Let's just do what you're doing here right now, it's gonna be fine.

**Review by petr-muller — APPROVED** [review]
/hold

LGTM, unhold if tests pass

**Pull request merged**
